### PR TITLE
added: support for a separate geometry in MxFiniteElement::Jacobian/Hessian

### PR DESCRIPTION
**Pull request by akva2**
Check for an additional entry at the end, and if so use this for the geometry basis.
Also adds Jacobian() for boundary use.

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this with downstreams please

**Comment by akva2** [comment]
jenkins build this with downstreams please

**Review by kmokstad — COMMENTED** [review]
The usage of the boundary-Jacobian method should also be included in this PR. With that, maybe the non-const `grad()` may indeed be protected.

File: src/ASM/FiniteElement.C [hunk 1]
```diff
@@ -110,22 +110,22 @@ bool MxFiniteElement::Hessian (Matrix3D& Hess, const Matrix& Jac,
...
-                      dNdX, false);
-
   size_t nBasis = bf ? bf->size() : d2Nxdu2->size();
-  for (size_t b = 2; b <= nBasis && ok; b++)
-    if (b != gBasis)
-      ok = utl::Hessian(Hess, d2MdX2[b-2], Jac, Xnod,
+  const bool separateGeometry = nBasis > this->getNoBasis();
+  bool ok;
+  if (separateGeometry)
+    ok = Hess.multiply(Xnod, (bf ? bf->back()->d2Ndu2 : d2Nxdu2->back()));
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/FiniteElement.C]: fixed.

File: src/ASM/FiniteElement.h [hunk 2]
```diff
@@ -51,6 +51,8 @@ class FiniteElement : public ItgPoint
 
   //! \brief Returns a const reference to the basis function 2nd-derivatives.
   virtual const Matrix3D& hess(char) const { return d2NdX2; }
+  //! \brief Returns a reference to the basis function 2nd-derivatives.
+  virtual Matrix3D& hess(char) { return d2NdX2; }
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/FiniteElement.h]: done.

**Comment by akva2** [comment]
making grad() protected will take some more work. it is used in some tests, and in jump integrands.

**Comment by akva2** [comment]
jenkins build this with downstreams please

**Comment by kmokstad** [comment]
It does not take that much actually, can be done like the fixup commit on top of this branch https://github.com/kmokstad/IFEM-2/tree/jacobian_hessian_separate_geom (not verified though).

**Comment by akva2** [comment]
jenkins build this with downstreams please

**Pull request merged**